### PR TITLE
add jnas recipe

### PR DESCRIPTION
**Pull request by Masao-Someki**
add jnas recipe.
this recipe was made under the following assumptions.
* structure of data directories is completely the same as that of
voxforge's.
* structure of .wav file names are completely the same as that of
vocforge'S.

### CODE REVIEW

File: egs/jnas/asr1/run.sh [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+maxlenratio=0.0
+minlenratio=0.0
+ctc_weight=0.3
+recog_model=model.acc.best # set a model to be used for decoding: 'model.acc.best' or 'model.loss.best'
+
+# scheduled sampling option
+samp_prob=0.0
+
+# data
+jnas=/database # original data directory to be stored
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: jnas_root=/database/JNAS

File: egs/jnas/asr1/run.sh [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+lang=ja # de, en, es, fr, it, nl, pt, ru
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: Remove.

File: egs/jnas/asr1/run.sh [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+### need to change
+train_set=train_nodup_sp
+train_dev=train_dev
+recog_set="eval1 eval2 eval3"
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: No need 3 recog sets. Just `eval` or `test` is OK.

File: egs/jnas/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+. ./cmd.sh
+
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+### need to change
+train_set=train_nodup_sp
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: train_set=train_nodev

File: egs/jnas/asr1/run.sh [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+train_dev=train_dev
+recog_set="eval1 eval2 eval3"
+###
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    ### Task dependent. You have to make data the following preparation part by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 0: Data Preparation"
+    selected=${jnas}/JNAS
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: Not needed.

File: egs/jnas/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    # Initial normalization of the data
+    local/jnas_data_prep.sh ${selected}
+    local/jnas_format_data.sh
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: You do not have to follow the voxforge.
If the processing is finished within the data_prep.sh, that is OK.

File: egs/jnas/asr1/run.sh [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+feat_dt_dir=${dumpdir}/${train_dev}/delta${do_delta}; mkdir -p ${feat_dt_dir}
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    ### Task dependent. You have to design training and dev sets by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 1: Feature Generation"
+    fbankdir=fbank
+    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
+    steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 10 --write_utt2num_frames true \
+        data/all_${lang} exp/make_fbank/train_${lang} ${fbankdir}
+    utils/fix_data_dir.sh data/all_${lang}
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: Should fix the names.
`data/train_*` or `data/eval`?

File: egs/jnas/asr1/run.sh [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d ${feat_tr_dir}/storage ]; then
+    utils/create_split_dir.pl \
+        /export/b{14,15,16,17}/${USER}/espnet-data/egs/jnas/asr1/dump/${train_set}/delta${do_delta}/storage \
+        ${feat_tr_dir}/storage
+    fi
+    if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d ${feat_dt_dir}/storage ]; then
+    utils/create_split_dir.pl \
+        /export/b{14,15,16,17}/${USER}/espnet-data/egs/jnas/asr1/dump/${train_dev}/delta${do_delta}/storage \
+        ${feat_dt_dir}/storage
+    fi
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]: remove L121-130.

File: egs/jnas/asr1/local/jnas_data_prep.sh [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+locdata=database/JNAS/local
+loctmp=$locdata/tmp
+rm -rf $loctmp >/dev/null 2>&1
+mkdir -p $locdata
+mkdir -p $loctmp
+
+# extract speakers names
+# from directory names into speakers_all.txt
+ls WAVES_HS | sort | sed -e "/@eaDir/d" > $loctmp/speakers_all.txt
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/jnas_data_prep.sh]: ls $wavdir?

File: egs/jnas/asr1/local/jnas_data_prep.sh [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+trans=Transcription
+type=KANJI
+wavdir=${DATA}/WAVES_HS
+ifNP=NP
+
+
+# make $DATA an absolute pathname.
+DATA=`perl -e '($dir,$pwd)= @ARGV; if($dir!~m:^/:) { $dir = "$pwd/$dir"; } print $dir; ' $DATA ${PWD}`
+
+locdata=database/JNAS/local
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/jnas_data_prep.sh]: You should not touch the database root directory.

File: egs/jnas/asr1/local/jnas_data_prep.sh [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+logdir=exp/jnas_data_prep
+mkdir -p $logdir
+echo -n > $logdir/make_trans.log
+for s in all; do
+    echo "--- Preparing ${s}_wav.scp, ${s}_trans.txt and ${s}.utt2spk ..."
+
+    for spkname in $(cat $loctmp/speakers_${s}.txt); do
+	scrdir=${DATA}/${trans}/${type}/${ifNP}
+  spkwav_dir=${wavdir}/${spkname}/${ifNP}
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/jnas_data_prep.sh]: fix indent.

File: egs/jnas/asr1/local/jnas_data_prep.sh [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	# check character set, and convert to utf-8
+	mkdir -p ${loctmp}/char_tmp
+	CHARSET=`file -bi ${scrdir}/${spkname}_${type:0:3}.txt |awk -F "=" '{print $2}'`
+	# echo -n $dir/etc/PROMPTS
+	# echo " encode:$CHARSET"
+	if [ "$CHARSET" != 'utf-8' ] && [ "$CHARSET" != 'us-ascii' ] ; then
+	    echo "iconv -f "$CHARSET" -t UTF-8 ${scrdir}/${spkname}_${type:0:3}.txt \
+      > ${loctmp}/char_tmp/$id.utf8"
+      iconv -f "$CHARSET" -t UTF-8 ${scrdir}/${spkname}_${type:0:3}.txt \
+      > ${loctmp}/char_tmp/$id.utf8
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/jnas_data_prep.sh]: fix indent.

**Comment by kan-bayashi** [comment]
Thank you for your quick response!
I’ve left some comment, could you fix them?

**Comment by Masao-Someki** [comment]
yes, of course!

**Comment by kan-bayashi** [comment]
Once finished, you should check whether the recipe could be finished without error.
And you can add the result as follows:
https://github.com/espnet/espnet/blob/master/egs/csj/asr1/RESULTS

**Comment by kan-bayashi** [comment]
I made PR to improve some parts.
Please check it.

**Comment by kan-bayashi** [comment]
BTW, it seems that text normalization is wrong.

```
                    "text": "カネ の 力 も 十分 知っ て い た",
                    "token": "カ ネ <space> の <space> 力 <space> も <space> 十 分 <space> 知 っ <space> て <space> い <space> た",
                    "tokenid": "96 128 2 51 2 379 2 71 2 408 352 2 1459 40 2 43 2 12 2 37"

```

In `text`, there are space, but it is not necessary.
This causes too bad result, I think.
You should remove space in the data preparation step.

**Comment by kan-bayashi** [comment]
Why don’t you make PR for espnet/espnet v.0.4.0?

**Comment by Masao-Someki** [comment]
now I'm testing whether this recipe works with good result.
After testing, I'll make PR to espnet/espnet.

**Comment by Masao-Someki** [comment]
I updated the result of jnas recipe.
It seems the results were going worse...

File: egs/jnas/asr1/run.sh [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+train_set=train_nodev
+train_dev=train_dev
+recog_set=(JNAS_testset_100 JNAS_testset_500)
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
recog_set="JNAS_testset_100 JNAS_testset_500"
```

File: egs/jnas/asr1/run.sh [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    ### Task dependent. You have to make data the following preparation part by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 0: Data Preparation"
+    # Initial normalization of the data
+    local/jnas_train_prep.sh ${jnas_root} ./local
+    for rtask in ${recog_set[@]};do
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
    for rtask in ${recog_set};do
```

File: egs/jnas/asr1/run.sh [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+      local/jnas_eval_prep.sh ${jnas_root}/DOCS/Test_set ${rtask}
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
        local/jnas_eval_prep.sh ${jnas_eval_root} ${rtask}
```

File: egs/jnas/asr1/local/train_speakers.txt [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+M148
+MP01
+MP02
+MP03
+MP04
+MP05
+
+
+
+
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/train_speakers.txt]: remove empty lines

File: egs/jnas/asr1/run.sh [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe) and Nagoya University (Someki Masao)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+. ./path.sh
+. ./cmd.sh
+
+# general configuration
+backend=pytorch
+stage=0       # start from -1 if you need to start from data download
+stop_stage=1
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
stop_stage=100
```

File: egs/jnas/asr1/run.sh [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+maxlenratio=0.0
+minlenratio=0.0
+ctc_weight=0.3
+recog_model=model.acc.best # set a model to be used for decoding: 'model.acc.best' or 'model.loss.best'
+
+# scheduled sampling option
+samp_prob=0.0
+
+# data
+jnas_root=/database/JNAS # original data directory to be stored
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
jnas_train_root=/database/JNAS               # root directory including original data
jnas_eval_root=/database/JNAS/DOCS/Test_set  # root directory including test data
```

File: egs/jnas/asr1/local/jnas_train_prep.sh [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+# extract speakers names
+
+logdir=exp/train_prep
+mkdir -p $logdir
+echo -n > $logdir/make_trans.log
+
+echo "--- Preparing train/wav.scp, train/trans.txt and train/utt2spk ..."
+
+for spkname in $(cat ${speaker_path}/train_speakers.txt); do
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/jnas_train_prep.sh]: spaker_path should be the path of speaker list, not directory.
```suggestion
for spkname in $(cat ${speaker_path}); do
```

File: egs/jnas/asr1/local/train_speakers.txt [hunk 20]
```diff
@@ -0,0 +1,265 @@
+F001
```
Comment (kan-bayashi) [inline egs/jnas/asr1/local/train_speakers.txt]: move to conf

File: egs/jnas/asr1/run.sh [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+train_set=train_nodev
+train_dev=train_dev
+recog_set=(JNAS_testset_100 JNAS_testset_500)
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    ### Task dependent. You have to make data the following preparation part by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 0: Data Preparation"
+    # Initial normalization of the data
+    local/jnas_train_prep.sh ${jnas_root} ./local
```
Comment (kan-bayashi) [inline egs/jnas/asr1/run.sh]:
```suggestion
    local/jnas_train_prep.sh ${jnas_train_root} conf/train_speakers.txt
```

**Comment by Masao-Someki** [comment]
I fixed all the typos and directories.

File: egs/jnas/asr1/run.sh (hunk 14 above)
Comment (kan-bayashi) [reply egs/jnas/asr1/run.sh]: @Masao-Someki Why don't you fix it?

**Comment by kan-bayashi** [comment]
Thanks! I'm testing now.

**Comment by Masao-Someki** [comment]
I'm sorry not to fixed the typo in run.sh
I just fixed it!

**Comment by kan-bayashi** [comment]
Good. Why don't you update RESULT using current recog_set?

**Comment by Masao-Someki** [comment]
It's already up-to-date!

**Comment by kan-bayashi** [comment]
Maybe you miss the result using JNAS_testset_500 due to the bug in run.sh (now is fixed).

**Comment by Masao-Someki** [comment]
ok. As soon as I get the result of this recipe, I'll update the result.

**Comment by Masao-Someki** [comment]
Now the results.md is up-to-date!

**Comment by kan-bayashi** [comment]
Great! 
LGTM

**Pull request merged**